ccontrol-1.0/ccontrol.1
'\" t
.\"     Title: ccontrol
.\"    Author: Rusty Russell <rusty@rustcorp.com.au>
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: v0.9 5 January 2006
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "CCONTROL" "1" "v0\&.9 5 January 2006" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccontrol \- wrapper to control distcc, ccache and more
.SH "SYNOPSIS"
.sp
\fIgcc\fR \&...
.sp
\fIcc\fR \&...
.sp
\fIc++\fR \&...
.sp
\fImake\fR \&...
.sp
\fIld\fR \&...
.sp
\fIccontrol\fR [\-\-section=<section>] <name> \&...
.sp
\fIccontrol\fR [\-\-section=<section>]
.SH "DESCRIPTION"
.sp
The ccontrol(1) program takes over the roles of the compiler and linker, and reads a configuration file to decide what to do before invoking them\&. This is particularly useful for centralized control over commands and options, such as enabling distcc(1) and ccache(1)\&.
.sp
When ccontrol(1) is invoked under its own name with no arguments, it prints out the settings which apply in this directory (unless \fI\-\-section\fR is specified)\&.
.sp
Versions are named after the last person to report a bug\&.
.SH "OPTIONS"
.sp
Normally ccontrol(1) is invoked as a symboling link to cc, make, etc, so it can identify what is being invoked by examining its own name\&. It can also be invoked under its own name, in which case ccontrol\-specific arguments can be supplied\&. The first non\-option argument will be used to identify the invocation, eg\&. "ccontrol gcc \&..."\&.
.sp
The following options are supported, when invoked as \fIccontrol\fR:
.PP
\-\-section=<section>
.RS 4
This is treated as the "current directory" for the purposes of evaluating the configuration file\&. As all real directories must begin with a "/" using an argument which does not, is a good way of overriding configuration for this particular invocation\&.
.RE
.SH "CONFIGURATION FILE"
.sp
ccontrol\(cqs configuration file is $HOME/\&.ccontrol/config\&. If this cannot be read (and written), your compilations will all fail\&. It is normal to have several different configuration files in this directory, and make default a symbolic link\&.
.SH "SYNTAX"
.sp
A configuration file consists of sections, led by a "[path]" header and followed by indented "name = value" entries\&. The first section is usually labelled "[*]" to set up the defaults\&. At the very least, you must set the "cc", "c++", "make" and "ld" values\&.
.sp
ccontrol will read every section which matches the current directory, so you can override values on a per\-directory basis\&. The "[path]" header of each section is a shell\-style wildcard (see glob(7)) which indicates the directory or directories it applies to\&. Usually this will end in a "*" to include all subdirectories\&.
.sp
All paths beginning with "~" are relative to the user\(cqs home directory\&. A path may be specified as a directory, in which case ccontrol will append the program name to the directory\&.
.sp
The following settings are available:
.PP
cc
.RS 4
Followed by
\fI=\fR
specifies the path of the compiler to be invoked when ccontrol is invoked as "cc" or "gcc"\&. ccontrol will fail to compile C programs if this is not set\&.
.RE
.PP
c++
.RS 4
Followed by
\fI=\fR
specifies the path of the compiler to be invoked when ccontrol is invoked as "c" or "g"\&. ccontrol will fail to compile C++ programs if this is not set\&.
.RE
.PP
ld
.RS 4
Followed by
\fI=\fR
specifies the path of the linker to be invoked when ccontrol is invoked as "ld"\&. ccontrol will fail to link programs if this is not set\&.
.RE
.PP
make
.RS 4
Followed by
\fI=\fR
specifies the path of the binary to be invoked when ccontrol is invoked as "make"\&. ccontrol will fail to make if this is not set\&.
.RE
.PP
ccache
.RS 4
Followed by
\fI=\fR
specifies the path of "ccache", and indicates that ccache is to be used where appropriate\&. If followed by
\fIdisable\fR, or not set, ccache will not be used\&.
.RE
.PP
distcc
.RS 4
Followed by
\fI=\fR
specifies the path of "distcc", and indicates that distcc is to be used where appropriate\&. If followed by
\fIdisable\fR, or not set, or distcc\-hosts is not set, distcc will not be used\&.
.RE
.PP
distcc\-hosts
.RS 4
Followed by
\fI=\fR
specifies the distcc servers to use, as per the DISTCC_HOSTS environment variable in distcc(1)\&. Followed by
\fIdisable\fR
disables distcc\&.
.RE
.PP
distc++\-hosts
.RS 4
The same as distcc\-hosts, but only applies to Ccompilations\&. If not set, distcc\-hosts is used\&. You can thus disable distcc for C
compilations by setting "distc++\-hosts disable"\&.
.RE
.PP
cpus
.RS 4
Followed by
\fI=\fR
and a number of CPUs, set to the number of CPUs you have (the default is "1")\&.
\fIccontrol\fR
uses this to tune the degree of parallelism\&.
.RE
.PP
no\-parallel
.RS 4
Followed by
\fI=\fR
and a space\-separated list of wildcards, suppresses parallel make for any make target matching one of those\&. This option is needed because ccontrol(1) usually forces make(1) to perform all actions in parallel, but this can be confusing when an error occurs, and breaks poorly\-written makefiles\&. Followed by
\fIdisable\fR, enables parallel make for all targets: this is useful to re\-enable parallel make in a subdirectory\&.
.RE
.PP
nice
.RS 4
Followed by
\fI=\fR
and a priority level from \-19 to 20, causes ccontrol to try to set its priority to this value\&. Default is 10\&.
.RE
.PP
include
.RS 4
Followed by
\fI=\fR
specifies a file to include at the current point\&. The effect is exactly as if the contents of the included file were literally inserted\&. Can be used at file level to include sections\&. Can also be used within sections to include section fragments\&.
.RE
.PP
add make
.RS 4
Followed by
\fI=\fR
specifies an argument to be added to each invocation of
\fImake\fR\&. This can be specified multiple times to add multiple arguments\&. Followed by
\fIdisable\fR
removes any arguments previously specified\&.
.RE
.PP
add env
.RS 4
Followed by
\fI=\fR
specifies an environment variable to be set, such as "add env = CCACHE_DIR=/tmp"\&. This can be specified multiple times to set multiple environment variables\&. Followed by
\fIdisable\fR
removes any arguments previously specified\&.
.RE
.PP
verbose
.RS 4
By itself, indicates that ccontrol(1) is to spit lots of crap out to standard error about what it\(cqs doing to your innocent command line\&.
.RE
.PP
lock\-file
.RS 4
Specify a particular lock file to use\&.
.RE
.SH "EXAMPLES"
.sp
This is the minimal configuration file:
.sp
.if n \{\
.RS 4
.\}
.nf
[*]
        cc = /usr/bin/gcc
        c++ = /usr/bin/g++
        ld = /usr/bin/ld
        make = /usr/bin/make
.fi
.if n \{\
.RE
.\}
.sp
If you have multiple locations (such as a laptop) it is common to have a "global" file which is included from every configuration file, like so:
.sp
.if n \{\
.RS 4
.\}
.nf
# Configuration file for when I\*(Aqm at work\&.  Lots of distcc hosts!
include = ~/\&.ccontrol/global

[*]
        distcc\-hosts = snab swarm1 swarm3 swarm4 swarm5 fandango2 mingo
        distc++\-hosts = snab mingo
.fi
.if n \{\
.RE
.\}
.sp
Here is a complete configuration file with several common scenarios:
.sp
.if n \{\
.RS 4
.\}
.nf
[*]
        cc = /usr/bin/gcc\-4\&.0
        c++ = /usr/bin/g++\-4\&.0
        ld = /usr/bin/ld
        make = /usr/bin/make
# Comment this back in for debugging
#       verbose
        distcc = /usr/bin/distcc
        distcc\-hosts = snab swarm1 swarm3 swarm4 swarm5 fandango2 mingo
        distc++\-hosts = snab mingo
        ccache = /usr/bin/ccache
        # make check should not generally be run in parallel
        no\-parallel = check

# Wesnoth doesn\*(Aqt compile with g++ 4\&.0
[*wesnoth*]
        c++ = /usr/bin/g++\-3\&.4

# Stupid third\-party modules don\*(Aqt build in parallel\&.
[/usr/src/modules/*]
        no\-parallel = *

# Using distcc when testing module\-init\-tools causes strange effects\&.
[*module\-init\-tools*/tests/*]
        distcc disable
.fi
.if n \{\
.RE
.\}
.SH "BUGS"
.sp
The ~/\&.ccontrol/config file must be writable: ccontrol(1) needs to get an exclusive write lock on it, which means it needs to open the file for writing\&. Use \fIinclude\fR to include read\-only files\&.
.sp
ccontrol will not immediately notice a change in included files, only in the toplevel file (ccontrol re\-reads the config file if it changed while ccontrol was trying to grab a lock)\&.
.sp
The Linux (\(la 2\&.6\&.15) cpufreq \fIondemand\fR governor (common on laptops) will not increase CPU speed when using ccontrol(1), because ccontrol re\-nices compilations\&. This can make builds 2\-3 times slower\&. Either use another governor, or tell \fIondemand\fR to ignore nice values:
.sp
.if n \{\
.RS 4
.\}
.nf
echo 1 > /sys/devices/system/cpu/cpu0/cpufreq/ondemand/ignore_nice
.fi
.if n \{\
.RE
.\}
.sp
If your code doesn\(cqt compile, ccontrol can only make it not compile faster\&.
.SH "AUTHOR"
.sp
Written by Rusty Russell <\m[blue]\fBrusty@rustcorp\&.com\&.au\fR\m[]\&\s-2\u[1]\d\s+2>
.SH "LICENSE"
.sp
Copyright \(co 2005 Rusty Russell\&. Free use of this software is granted under the terms of the GNU General Public License (GPL)\&.
.SH "SEE ALSO"
.sp
make(1), cc(1), c++(1), ld(1), distcc(1), ccache(1), glob(7), cpufreq\-set(1)
.SH "AUTHOR"
.PP
\fBRusty Russell\fR <\&rusty@rustcorp\&.com\&.au\&>
.RS 4
Author.
.RE
.SH "NOTES"
.IP " 1." 4
rusty@rustcorp.com.au
.RS 4
\%mailto:rusty@rustcorp.com.au
.RE


ccontrol-1.0/CHANGES
Changes for 1.0
---------------
Configuration language changes:
	- config file now called ~/.ccontrol/config not default
	- '= off' now replaced with 'disable', can be used in more places
	- 'parallel' keyword gone: use 'no-parallel disable'
	- 'no-parallel' by itself no longer valid, use 'no-parallel = *'
	- 'nice' keyword to change priority
	- 'lock-file' keyword to have some builds not block others.
	- Paths (include, distcc, etc) expand ~ so you can use "~/bin/gcc".
	- Unknown argv[0] is passed through, so you can ccontrol anything.
	

Changes for 0.9
---------------
New features:
	- Gnome panel control-area applet which sets symlink

Other changes:
	- Fix PowerPC (and other?) locking (patch from Tony Breeds)
	- Intercept ccache misses, so we don't have to guess chance of hit
	- Fix error reporting (also patched by Olof Johansson)

Changes for 0.8
---------------
Other changes:
	- fix bug where all complations done at once if distcc failed.
	- invoking ccontrol by itself gives current settings.
	- invoking ccontrol -V/--version gives current version.
	- debian packaging
	- fix bug where ccontrol didn't reread config file when re-symlinked

Changes for 0.7
---------------
Configuration language changes:
	- "add make" to add makefile arguments (eg. for cross-compiling).

Other changes:
	- ccontrol can now be invoked with --section= for special effects.
	- Faster response to config changes if distcc fails.
	- Examples and troubleshooting in documentation

Changes for 0.6
---------------
Configuration language changes:
	- distc++-hosts for c++ compiles (defaults to distcc-hosts)
	- parallel keyword to override no-parallel
	- distcc only used if distcc-hosts/distc++-hosts set

Other changes:
	- toplevel make now shouldn't queue behind other makes.

Changes for 0.5
---------------
Configuration language changes:
	- off keyword for distcc and ccache to turn them off locally

Other changes:
	- ccontrol: "cannot decrement semaphore" fixed: support multiple users.
	- distcc: detect when distcc is failing, don't flood localhost.
	- documentation: fix cpus entry in ccontrol(1)
	- ccontrol-init: human names, fix some failure cases
	- ccache: don't limit distcc so hard when ccache enabled
	- install: create directories if necessary

Changes for 0.4
---------------
IMPORTANT: Configuration language changes:

	- "parallel" renamed to "cpus" (other was confusing)

	- "[global]" must be replaced by "[*]": we now process *every*
	  matching section, in order.

Other changes:
	ccontrol-init: better handling for probing distcc hosts
	configuration: "include" keyword for including files
	documentation: clarifications, hacking advice, license
	distribution: include .hg dir to help users become developers
	users: ccontrol now has a second user (and developer)
	


ccontrol-1.0/ccontrol.c
#define _GNU_SOURCE
#include <limits.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fnmatch.h>
#include <stdarg.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <fcntl.h>
#include "stdrusty.h"
#include "ccontrol.h"

static void insert_arg(char *argv[], unsigned argc, unsigned pos, char *arg)
{
	memmove(argv+pos+1, argv+pos, (argc+1) * sizeof(char *));
	argv[pos] = arg;
}

static void write_string(int fd, const char *str)
{
	write(fd, str, strlen(str));
}

void __verbose(const char *msg, ...)
{
	va_list arglist;
	const char *str;

	write_string(STDERR_FILENO, "   ccontrol(");
	write_string(STDERR_FILENO, int_to_string(getpid()));
	write_string(STDERR_FILENO, ") ");
	write_string(STDERR_FILENO, msg);

	va_start(arglist, msg);
	while ((str = va_arg(arglist, char *)) != NULL)
		write_string(STDERR_FILENO, str);
	va_end(arglist);

	write_string(STDERR_FILENO, "\n");
}

static bool make_target_match(char *argv[], char *targets)
{
	char *p, *target = "";
	unsigned int i;

	if (!targets)
		return false;

	if (!targets[0])
		return true;

	/* Heuristic: choose last arg which is neither option nor variable */
	for (i = 0; argv[i]; i++) {
		if (argv[i][0] == '-')
			continue;
		if (strchr(argv[i], '='))
			continue;
		target = argv[i];
	}

	for (p = strtok(targets, " \t"); p; p = strtok(NULL, " \t"))
		if (fnmatch(p, target, 0) == 0)
			return true;
	return false;
}

static bool already_has_dash_j(char *argv[])
{
	unsigned int i;
	const char *makeflags;

	for (i = 0; argv[i]; i++)
		if (strstarts(argv[i], "-j"))
			return true;

	/* Top-level -j<n> sets --jobserver-fds for kids. */
	makeflags = getenv("MAKEFLAGS");
	return makeflags && strstr(makeflags, "--jobserver-fds") != NULL;
}

/* Earlier arg parsing might have revealed distcc is useless. */
static void adjust_args_and_environment(struct section sec,
					enum type type,
					char *argv[], unsigned argc)
{
	struct add *add;

	if (type != TYPE_OTHER)
		argv[0] = sec.names[type];

	switch (type) {
	case TYPE_MAKE:
		/* Ensure make calls us for children, not /usr/bin/make. */
		setenv("MAKE", "make", 1);

		/* We insert -j to make in parallel. */
		if (!make_target_match(argv, sec.no_parallel)
		    && !already_has_dash_j(argv)) {
			const char *cpus = int_to_string(sec.cpus*20);
			char *arg = malloc(strlen(cpus) + 3);

			strcpy(arg, "-j");
			strcpy(arg + strlen(arg), cpus);
			insert_arg(argv, argc++, 1, arg);
		}
		while (sec.make_add) {
			insert_arg(argv, argc++, 1, sec.make_add->arg);
			sec.make_add = sec.make_add->next;
		}
		break;
	case TYPE_CC:
	case TYPE_CPLUSPLUS:
		if (sec.ccache) {
			verbose(sec.verbose, "Prefixing ccache ", sec.ccache);
			insert_arg(argv, argc++, 0, sec.ccache);
		} else if (sec.distcc) {
			char *hosts;

			if (type == TYPE_CC)
				hosts = sec.distcc_hosts;
			else
				hosts = sec.distcplusplus_hosts;
			verbose(sec.verbose, "Setting DISTCC_HOSTS=", hosts);
			setenv("DISTCC_HOSTS", hosts, 1);
			verbose(sec.verbose, "Prefixing arg ", sec.distcc);
			insert_arg(argv, argc++, 0, sec.distcc);
		}

	case TYPE_LD:
	case TYPE_OTHER:
		break;
	}

	for (add = sec.env_add; add; add = add->next) {
		char *eq = strchr(add->arg, '=');
		if (!eq)
			fatal("Environment variable needs '=': ", 0, add->arg,
			      NULL);
		*eq = '\0';
		setenv(add->arg, eq+1, 1);
	}
}

#ifndef TESTING
static void __attribute__((noreturn)) run_command(bool verbose, char *argv[])
{
	if (verbose) {
		unsigned int i;
		write_string(STDERR_FILENO, "  ccontrol(");
		write_string(STDERR_FILENO, int_to_string(getpid()));
		write_string(STDERR_FILENO, ") execing ");
		for (i = 0; argv[i]; i++) {
			write_string(STDERR_FILENO, "'");
			write_string(STDERR_FILENO, argv[i]);
			write_string(STDERR_FILENO, "' ");
		}
		write_string(STDERR_FILENO, "\n");
	}
	execv(argv[0], argv);
	fatal("failed to exec '", errno, argv[0], "'", NULL);
}
#else
static void __attribute__((noreturn)) run_command(bool verbose, char *argv[])
{
	/* Print out all TESTING= env vars. */
	if (getenv("TESTING1"))
		printf("TESTING1='%s' ", getenv("TESTING1"));
	if (getenv("TESTING2"))
		printf("TESTING2='%s' ", getenv("TESTING2"));

	if (strcmp(getenv("FAILDISTCC") ?: "", "1") == 0)
		exit(103);

	if (getenv("DISTCC_HOSTS")) {
		printf("DISTCC_HOSTS='%s' ", getenv("DISTCC_HOSTS"));
	}

	while (argv[0]) {
		printf("%s ", argv[0]);
		argv++;
	}
	printf("\n");

	exit(atoi(getenv("EXITCODE") ?: "0"));
}
#endif

/* Fork command. */
static void fork_command(bool verbose, char *argv[])
{
	int pid = fork();

	if (pid < 0)
		fatal("failed to fork", errno, NULL);
	if (pid == 0)
		run_command(verbose, argv);
	verbose(verbose, "Forked ", int_to_string(pid));
}

/* Wait for forked command */
static int wait_for_child(bool verbose)
{
	int status;

	wait(&status);
	status = WIFEXITED(status) ? WEXITSTATUS(status) : 255;
	verbose(verbose, "Child returned ", int_to_string(status));
	return status;
}

static bool file_altered(const char *configname, const struct stat *st)
{
	struct stat st2;

	stat(configname, &st2);
	return st2.st_ctime != st->st_ctime || st2.st_mtime != st->st_mtime
		|| st2.st_ino != st->st_ino;
}

static void write_item(const char *a, const char *b)
{
	if (b) {
		write_string(STDOUT_FILENO, "\t");
		write_string(STDOUT_FILENO, a); write_string(STDOUT_FILENO, b);
		write_string(STDOUT_FILENO, "\n");
	}
}

/* This list is backwards.  Print it forwards. */
static void print_add(const char *str, struct add *add)
{
	if (!add)
		return;
	print_add(str, add->next);
	write_item(str, add->arg);
}

static void print_section(const char *dir, struct section sec)
{
	write_string(STDOUT_FILENO, "[");
	write_string(STDOUT_FILENO, dir);
	write_string(STDOUT_FILENO, "]\n");
	write_item("cc = ", sec.names[TYPE_CC]);
	write_item("c++ = ", sec.names[TYPE_CPLUSPLUS]);
	write_item("make = ", sec.names[TYPE_MAKE]);
	write_item("ld = ", sec.names[TYPE_LD]);
	write_item("ccache = ", sec.ccache);
	if (sec.distcc && (sec.distcc_hosts || sec.distcplusplus_hosts)) {
		write_item("distcc = ", sec.distcc);
		write_item("distcc-hosts = ", sec.distcc_hosts);
		if (sec.distcplusplus_hosts_set)
			write_item("distc++-hosts = ",sec.distcplusplus_hosts);
	}
	print_add("add make = ", sec.make_add);
	print_add("add env = ", sec.env_add);
	if (sec.no_parallel && streq(sec.no_parallel, ""))
		write_item("no-parallel", "");
	else
		write_item("no-parallel = ", sec.no_parallel);
	if (sec.cpus != 1)
		write_item("cpus = ", int_to_string(sec.cpus));
	if (sec.verbose)
		write_item("verbose", "");
	if (sec.nice != 10)
		write_item("nice = ", int_to_string(sec.nice));
	write_item("lock-file = ", sec.lock_file);
}

static bool is_preprocess(char *argv[])
{
	char **arg;
	for (arg = argv; *arg; arg++) {
		if (streq(*arg, "-E"))
			return true;
	}
	return false;
}

int main(int orig_argc, char *orig_argv[])
{
	enum type type;
	struct section sec;
	int fd, ret, argc;
	bool nodistcc = false, noccache = false, ccache_then_distcc = false;
	struct stat st;
	char configname[PATH_MAX];
	char dirname[PATH_MAX];
	char *new_argv[orig_argc + 23], **argv;
	undofn_t undo;

	/* Have we already done ccache? */
	if (getenv("CCONTROL_LOCK_FD")) {
		/* Ccache calls us once to pre-process. */
		if (is_preprocess(orig_argv+1))
			run_command(false, orig_argv+1);
		/* Write to fd to make grandparent (ccontrol) drop lock */
		if (write(atoi(getenv("CCONTROL_LOCK_FD")), "#", 1) != 1)
			fatal("Could not write to ccontrol lock fd",
			      errno, NULL);
		unsetenv("CCACHE_PREFIX");
		unsetenv("CCONTROL_LOCK_FD");
		noccache = true;
	}

	strcpy(configname, getenv("HOME"));
	strcat(configname, "/.ccontrol/config");

again_restore_args:
	/* Make room to add args. */
	argc = orig_argc;
	memcpy(new_argv, orig_argv, (argc + 1)*sizeof(argv[0]));
	argv = new_argv;

#ifdef TESTING
	if (getenv("ARGV0"))
		argv[0] = getenv("ARGV0");
#endif
	getcwd(dirname, sizeof(dirname));

	if (strends(argv[0], "ccontrol")) {
		if (argv[1] && strstarts(argv[1], "--section=")) {
			strcpy(dirname, argv[1] + strlen("--section="));
			argv++;
			argc--;
		}
		if (argv[1] && (streq(argv[1], "--version")
				|| streq(argv[1], "-V"))) {
			
			__verbose("version " VERSION, NULL);
			exit(0);
		}
		if (!argv[1]) {
			print_section(dirname,
				      read_config(configname, dirname,
						  open(configname, O_RDWR)));
			exit(0);
		}
		argv++;
		argc--;
	}
	type = what_am_i(argv);

again:
	/* Since we later grab an exclusive lock on this, must be writable */
	fd = open(configname, O_RDWR);

	/* This handles open failure if fd < 0. */
	sec = read_config(configname, dirname, fd);

    /* Ensure we exec the right command */
    {
        char *cmd = resolve_path(sec.names[type], argv[0]);
        free(sec.names[type]);
        sec.names[type] = cmd;
    }

	fstat(fd, &st);

	/* Run low priority; people like to use apps while compiling. */
	setpriority(PRIO_PROCESS, 0, sec.nice);

	if (noccache) {
		verbose(sec.verbose, "ccache suppressed this time");
		sec.ccache = NULL;
	}

	/* Check we really distcc if asked to. */
	if (sec.distcc) {
		if (nodistcc || type == TYPE_LD || type == TYPE_MAKE)
			sec.distcc = NULL;
		else if (sec.ccache) {
			/* Don't put this in command line, but see below */
			sec.distcc = NULL;
			ccache_then_distcc = true;
		} else {
			if (type == TYPE_CC && !sec.distcc_hosts)
				sec.distcc = NULL;
			else if (type == TYPE_CPLUSPLUS
				 && !sec.distcplusplus_hosts)
				sec.distcc = NULL;
			else if (!can_distcc(argv)) {
				verbose(sec.verbose, "Cannot distcc this");
				sec.distcc = NULL;
			}
		}
	}

	/* Grab lock returns undofn if it slept: restat config file.  This
	 * decreases latency when config file changed. */
	undo = grab_lock(sec.lock_fd, sec, type);
	if (undo && file_altered(configname, &st)) {
		/* If we had distcc disabled, try again. */
		nodistcc = 0;
		close(sec.lock_fd);
		undo();
		goto again;
	}

	adjust_args_and_environment(sec, type, argv, argc);

	if (ccache_then_distcc) {
		/* We need to hold slow lock until either ccache pulls
		   from cache and exits, OR ccontrol is invoked from
		   ccache (cache miss).
		 */
		char c;
		int fd[2];

		if (pipe(fd) != 0)
			fatal("Failed to create pipe", errno, NULL);
		verbose(sec.verbose, "Setting CCACHE_PREFIX=ccontrol");
		setenv("CCONTROL_LOCK_FD", int_to_string(fd[1]), 1);
		setenv("CCACHE_PREFIX", "ccontrol", 1);
		fork_command(sec.verbose, argv);
		close(fd[1]);
		/* This will return when child exits, OR writes manually. */
		if (read(fd[0], &c, 1) == 1)
			verbose(sec.verbose, "Grandchild said to drop lock");
		else
			verbose(sec.verbose, "I guess ccache exited");
		undo();
		exit(wait_for_child(sec.verbose));
	}

	if (!sec.distcc)
		run_command(sec.verbose, argv);

	/* We have to wait, to make sure distcc actually distributes,
	 * otherwise we end up running 20 compiles locally. */
	setenv("DISTCC_FALLBACK", "0", 1);

	fork_command(sec.verbose, argv);
	ret = wait_for_child(sec.verbose);
	if (ret == 103 || ret == 116 || ret == 107) {
		/* distcc failed to distribute: run single-thread. */
		verbose(sec.verbose, "distcc failed, retrying without");
		close(sec.lock_fd);
		undo();
		nodistcc = 1;
		unsetenv("CCACHE_PREFIX");
#ifdef TESTING
		unsetenv("FAILDISTCC");
#endif
		while (sec.env_add) {
			unsetenv(sec.env_add->arg);
			sec.env_add = sec.env_add->next;
		}
		goto again_restore_args;
	}
	return ret;
}
